### ISSUE TITLE
firefox fails to build on release-17.09

### ISSUE THREAD

**Issue by basvandijk** [NixOS/nixpkgs#31706]
Building firefox fails on the latest `release-17.09` (b0e50cc7b97f96c7624bcabebf9ae5de51566225):

```
$ nix-build -A firefox
...
ERROR: Rust compiler 1.17.0 is too old.

To compile Rust language sources please install at least
version 1.19.0 of the 'rustc' toolchain and make sure it is
first in your path.
```

We probably need to cherry-pick the following on `release-17.09`:

commit 7bd191df6a861e02d5bf2b36a03f6dc70e3e6bfd
Author: zimbatm <zimbatm@zimbatm.com>
Date:   Wed Oct 4 17:41:16 2017 +0100

    rust: 1.17.0 -> 1.20.0

commit d170c2ceadde34a1ee4844fe59d876351331b707
Author: zimbatm <zimbatm@zimbatm.com>
Date:   Sun Oct 8 19:26:40 2017 +0100

    cargo: use fetchFromGitHub

commit 5f8cf0048ea089fa73d17512fc4f9f0f0644e225
Author: Kevin Cox <kevincox@kevincox.ca>
Date:   Sat Aug 5 15:38:48 2017 +0100

    rust: update cargo builder to fetch registry dynamically

commit 6bbce42fe7909fc0096f27be2f7958cf53f6ac17
Author: zimbatm <zimbatm@zimbatm.com>
Date:   Thu Oct 19 22:29:28 2017 +0000

    cargo: 0.21.1 -> 0.22.0

Unfortunately they can't be cherry-picked cleanly.

@zimbatm @taku0 ping.

**Comment by zimbatm** [comment]
are we allowing to change major versions of firefox in stable branches? upgrading rust is not completely without it's own side-effects.

**Comment by NeQuissimus** [comment]
Well, FF changes "major versions" with virtually every other release :D

**Comment by vcunat** [comment]
I suppose we'll have to keep two rusts on stable, if we want to keep secure non-ESR FF in there and not upgrading rust for all _other_ use cases as well.  /cc 17.09 managers: @globin, @fpletz.

**Comment by grahamc** [comment]
We essentially must backport all browser updates back to stable, we don't have the man power to backport just patches.

**Comment by fpletz** [comment]
The nss & nspr updates were not backported either which are presumably also needed.

**Comment by edolstra** [comment]
@zimbatm Yes. People who don't want major Firefox changes should use firefox-esr.

Changing rust, OTOH, is probably not acceptable for the stable branch.

**Comment by domenkozar** [comment]
Uhh. Sorry about this. Should we have a separate rust then?

**Comment by edolstra** [comment]
Yes, sounds good.

**Comment by fpletz** [comment]
@domenkozar I've reverted your firefox bump in 5e3cfdeb6228ae03e30bd331cd6b676cb7018536 on 17.09 for now because it blocks channel releases. The ACME fix from #31745 has to be shipped asap.

**Comment by basvandijk** [comment]
Thanks guys.
